### PR TITLE
Documentation: Add the documentation for assignment 3

### PR DESCRIPTION
**Pull request by dbaluta**
Add SO2 assigment 3. Original is at:
https://ocw.cs.pub.ro/courses/so2/teme/tema3

The assignment requests to create a software RAID1 kernel module.

Signed-off-by: Daniel Baluta <daniel.baluta@nxp.com>

### CODE REVIEW

**Comment by dbaluta** [comment]
@StefanTeodorescu initial draft need to fix links, add images.

File: Documentation/teaching/so2/assign3-software-raid.rst [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+- the `make_request_fn` function can be called in parallel
+- since data sectors are separated from CRC sectors you will have to build separate `bio` structures for data and CRC values
+- to allocate a ``bio`` for physical disks you can use ``bio_alloc``; to add data pages to bio use ``alloc_page`` and ``bio_add_page``
+- to free up the space allocated by a `bio` you need to release the pages allocated to the bio (using the ``__free_page`` macro ) and call ``bio_put``
+- when generating a ``bio`` structure, consider that it must be multiple times the size of the disk sector (``KERNEL_SECTOR_SIZE``)
+- to send a request to a block device and wait for it to end, you can use the ``submit_bio_wait`` function
+- use ``bio_endio`` to signal the completion of processing a bio structure
+- for the CRC32 calculation you can use the ``crc32`` macro provided by the kernel
+- useful macro definitions can be found in the assignment support `header <http://elf.cs.pub.ro/so2/res/teme/ssr.h>`__
+- our recommendations are not manddatory (any solution that meets the requirements of the assignment are accepted)
```
Comment (StefanTeodorescu) [inline Documentation/teaching/so2/assign3-software-raid.rst:72]: manddatory -> mandatory

are accepted -> is accepted

File: Documentation/teaching/so2/assign3-software-raid.rst [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+
+- A single request processing function for block devices can be active at one time in a call stack (more details `here <https://elixir.bootlin.com/linux/v5.10/source/block/blk-core.c#L1048>`__).
+You will need to submit requests for physical devices in a thread kernel; we recommend using ``workqueues``.
```
Comment (StefanTeodorescu) [inline Documentation/teaching/so2/assign3-software-raid.rst:75]: thread kernel -> kernel thread

File: Documentation/teaching/so2/assign3-software-raid.rst [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+- For a quick run, use a single bio to batch send the read/write request for CRC values for adjacent sectors. For example,
+if you need to send requests for CRCs in sectors 0, 1, 7, use a single bio, not 8 bios.
```
Comment (StefanTeodorescu) [inline Documentation/teaching/so2/assign3-software-raid.rst:77]: 0, 1, ..., 7

**Review by StefanTeodorescu — APPROVED** [review]

File: Documentation/teaching/so2/assign3-software-raid.rst [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+Implementation Details
+======================
+
+- the kernel module will be named ``ssr.ko``
+- the logical device will be accessed as a block device with the major ``SSR_MAJOR`` and minor ``SSR_FIRST_MINOR`` under the name ``/dev/ssr`` (via the macro ``LOGICAL_DISK_NAME``)
+- the virtual device (``LOGICAL_DISK_NAME``-``/dev/ssr``) will have the capacity of ``LOGICAL_DISK_SECTORS`` (use ``set_capacity`` with the ``struct gendisk`` structure)
+- the two disks are represented by the devices ``/dev/vdb``, respectively ``/dev/vdc``, defined by means of macros ``PHYSICAL_DISK1_NAME``, respectively ``PHYSICAL_DISK2_NAME``
+- to work with the ``struct block _device`` structure associated with a physical device, you can use the ``blkdev_get_by_path`` and ``blkdev_put`` functions
+- for the handling of requests from the user space, we recommend not to use a queue of requests, but to do processing at the level of struct ``bio`` structures
+  you can register an appropriate routine using the ``blk_queue_make_request`` call
```
[the remark this answers is not in the document]
Comment (dbaluta) [reply Documentation/teaching/so2/assign3-software-raid.rst:62]: corect. am rezolvat.

File: Documentation/teaching/so2/assign3-software-raid.rst (hunk 2 above)
Comment (dbaluta) [reply Documentation/teaching/so2/assign3-software-raid.rst:75]: Fixed

**Pull request merged**